Makefile

```diff
@@ -7,195 +7,164 @@ PG_VERSIONS?=12 11
+#  3. timescale/timescaledb-ha:pg12.3-ts1.7-latest
+#  4. timescale/timescaledb-ha:pg12.3-ts1.7.1-latest
+#  5. timescale/timescaledb-ha:pg12.3-ts1.7.1-pN
 #
 # Tag 5 is immutable, and for every time we publish that image, we increase N by 1,
 # we start with N=0
-#
-# Our method of finding a patch version is quite brute force (`docker pull image`),
-# however, this should not happen that often.
-.PHONY: publish publish-oss
-publish publish-oss:
-	export PGVERSION=$$(docker inspect $(CICDIMAGE) | jq '.[0]."ContainerConfig"."Labels"."com.timescaledb.image.postgresql.version"' -r) \
-	&& export TSPATCH=$$(docker inspect $(CICDIMAGE) | jq '.[0]."ContainerConfig"."Labels"."com.timescaledb.image.timescaledb.version"' -r) \
-	&& export TSMINOR=$${TSPATCH%.*} \
-	&& for variant in pg$(PG_MAJOR)$(POSTFIX)-latest pg$(PG_MAJOR)-ts$${TSMINOR}$(POSTFIX)-latest pg$${PGVERSION}-ts$${TSMINOR}$(POSTFIX)-latest pg$${PGVERSION}-ts$${TSPATCH}$(POSTFIX)-latest; \
-		do \
-			docker tag $(CICDIMAGE) $(PUBLISH_REPOSITORY):$${variant} \
-			&& docker push $(PUBLISH_REPOSITORY):$${variant} || exit 1; \
+publish: publish-mutable publish-next-patch-version
+
+is_ci:
+	@if [ "$${CI}" != "true" ]; then echo "environment variable CI is not set to \"true\", are you running this in Github Actions?"; exit 1; fi
+
+publish-mutable: is_ci build
+	for latest in pg$(PG_MAJOR) pg$(PG_MAJOR)-ts$(VAR_TSMAJOR) pg$(VAR_PGMINOR)-ts$(VAR_TSMAJOR) pg$(VAR_PGMINOR)-ts$(VAR_TSMINOR); do \
+		for url in $(DOCKER_PUBLISH_URLS); do \
+			docker tag $(DOCKER_TAG_LABELED) $${url}:$${latest}$(DOCKER_TAG_POSTFIX)-latest || exit 1; \
+			docker push $${url}:$${latest}$(DOCKER_TAG_POSTFIX)-latest || exit 1 ; \
 		done \
-	&& for i in $$(seq 0 100); do \
-		export IMMUTABLE_TAG=$(PUBLISH_REPOSITORY):pg$${PGVERSION}-ts$${TSPATCH}$(POSTFIX)-p$${i} \
-		&& if docker pull $${IMMUTABLE_TAG}; \
-			then \
-				echo "$${IMMUTABLE_TAG} already exists"; \
-			else \
-				docker tag $(CICDIMAGE) $${IMMUTABLE_TAG}; \
-				docker push $${IMMUTABLE_TAG} && exit 0 || exit 1; \
-			fi; \
-	   done
-
-.PHONY: publish
-publish-all:
-ifndef RELEASE_TAG
-	$(error RELEASE_TAG is undefined, please set it to a tag that was succesfully built)
+	done
+
+
+publish-immutable: is_ci build
+	for i in $$(seq 0 100); do \
+		export IMMUTABLE_TAG=pg$(VAR_PGMINOR)-ts$(VAR_TSMINOR)$(DOCKER_TAG_POSTFIX)-p$${i}; \
+		export DOCKER_HUB_HTTP_CODE="$$(curl -s -o /dev/null -w '%{http_code}' "$(DOCKER_CANONICAL_URL)/tags/$${IMMUTABLE_TAG}")"; \
+		if [ "$${DOCKER_HUB_HTTP_CODE}" = "404" ]; then \
+			for url in $(DOCKER_PUBLISH_URLS); do \
+				docker tag $(DOCKER_TAG_LABELED) $${url}:$${IMMUTABLE_TAG} || exit 1; \
+				docker push $${url}:$${IMMUTABLE_TAG} || exit 1 ; \
+			done; \
+			exit 0; \
+		elif [ "$${DOCKER_HUB_HTTP_CODE}" = "200" ]; then \
+			echo "$${IMMUTABLE_TAG} already exists, incrementing patch number"; \
+		else \
+			echo "Unexpected HTTP return code: $${DOCKER_HUB_HTTP_CODE}"; \
+			exit 1 ;\
+		fi \
+	done
+
```

Comment:
very unlikely to happen, but should we complain if we reached past 100 without coming across non-existing image?

Reply:
Yeah, let's add it. > 100 would be a bug probably, but better to complain than to succeed silently.